File: day01.py
import os,math, difflib, pathlib, string, re, json, functools, networkx as nx, numpy as np, heapq
from itertools import permutations, combinations, chain, cycle, repeat, product
from collections import defaultdict, Counter, deque, ChainMap
from copy import deepcopy

cur_dir = pathlib.Path(__file__).resolve().parents[0]
file_name ="/input01.txt"
lines = open(str(cur_dir)+file_name).read().split("\n")


ans = 0

for x in lines:
    digits = [c for c in x if c.isdigit()]
    ans += int(digits[0] + digits[-1])

print("part1:", ans)



ans = 0
digits = "one two three four five six seven eight nine".split()

d2n = dict((digit, str(i)) for digit, i in zip(digits, range(1,10)))
regExp = "(?=(" + "|".join(digits) + "|\\d))"

def help(x):
    if x in d2n:
        return d2n[x]
    return x

for line in lines:
    digits = [*map(help, re.findall(regExp, line))]
    ans += int(digits[0] + digits[-1])

print("part2:", ans)

File: day01.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"unicode"
)

func main() {	
	data, err := os.ReadFile("input01.txt")
	if err != nil {
		panic("Cannot read input file")
	}
	lines := strings.Split(string(data), "\n")	
	part1(lines)
	part2(lines)
}	

func part1(lines []string){	
	res := 0
	for _, line := range lines {
		var digits []rune
		for _, digit := range line {
			if unicode.IsDigit(digit) {
				digits = append(digits, digit)				
			}
		}
		sum := string(digits[0]) + string(digits[len(digits)-1])
		invValue, err := strconv.Atoi(sum)
		if err != nil {
			panic("Cannot convert to int")
		}
		res += invValue
	}

	fmt.Println("part1:", res)	
}

func part2( lines []string){
	sum := 0
	for _, line := range lines {	
		sum +=processLine(line)
	}
	fmt.Println("part2:", sum)	
}

func processLine(input string)int {
	// Mapping of number words to their numeric values
	numberWords := map[string]string{
		"zero": "0", "one": "1", "two": "2", "three": "3", "four": "4",
		"five": "5", "six": "6", "seven": "7", "eight": "8", "nine": "9",
	}

	output := ""
	i := 0

	for i < len(input) {		
		matched := false
		for word, value := range numberWords {
			if strings.HasPrefix(input[i:], word) {
				output += value
				matched = true
			}
		}	
		if !matched{
			if input[i] >= '0' && input[i] <= '9' {
				output += string(input[i])
				
			}
		}
		i++
	}
	res, err := strconv.Atoi(string(output[0]) + string(output[len(output)-1]))
	if err != nil {	
		panic("Cannot convert to int")	
	}
	return res
}